packages/compiler-cli/src/ngtsc/transform/src/implicit_signal_debug_name_transform.ts

```diff
@@ -0,0 +1,473 @@
+    const nodeText = node.name.getText();
+
+    return ts.factory.updateVariableDeclaration(
+      node,
+      node.name,
+      node.exclamationToken,
+      node.type,
+      insertDebugNameIntoCallExpression(node.initializer, nodeText),
+    );
+  } catch {
+    return node;
+  }
+}
+
+function transformPropertyAssignment(
+  program: ts.Program,
+  node: ts.ExpressionStatement & {
+    expression: ts.BinaryExpression & {right: ts.CallExpression; left: ts.PropertyAccessExpression};
+  },
+): ts.ExpressionStatement {
+  const expression = node.expression.right.expression;
+  if (ts.isPropertyAccessExpression(expression)) {
+    if (!expressionIsUsingAngularCoreImportedSymbol(program, expression.expression)) {
+      return node;
+    }
+  } else if (!expressionIsUsingAngularCoreImportedSymbol(program, expression)) {
+    return node;
+  }
+
+  return ts.factory.updateExpressionStatement(
+    node,
+    ts.factory.createBinaryExpression(
+      node.expression.left,
+      node.expression.operatorToken,
+      insertDebugNameIntoCallExpression(node.expression.right, node.expression.left.name.text),
+    ),
+  );
+}
+
+function transformPropertyDeclaration(
+  program: ts.Program,
+  node: ts.PropertyDeclaration,
+): ts.PropertyDeclaration {
+  if (!node.initializer || !ts.isCallExpression(node.initializer)) return node;
+
+  const expression = node.initializer.expression;
+  if (ts.isPropertyAccessExpression(expression)) {
+    if (!expressionIsUsingAngularCoreImportedSymbol(program, expression.expression)) {
+      return node;
+    }
+  } else if (!expressionIsUsingAngularCoreImportedSymbol(program, expression)) {
+    return node;
+  }
+
+  try {
+    // may throw if the node does not have a source file. Ignore this case for now.
+    const nodeText = node.name.getText();
+    return ts.factory.updatePropertyDeclaration(
+      node,
+      node.modifiers,
+      node.name,
+      node.questionToken,
+      node.type,
+      insertDebugNameIntoCallExpression(node.initializer, nodeText),
+    );
+  } catch {
+    return node;
+  }
+}
+
+/**
+ *
+ * This transformer adds a debugName property to the config object of signal functions like
+ * signal, computed, effect, etc.
+ *
+ * The debugName property is added conditionally based on the value of ngDevMode. This is done
+ * to avoid adding the debugName property in production builds.
+ *
+ * Ex:
+ * ```ts
+ * import {signal} from '@angular/core';
+ * const mySignal = signal('Hello World');
+ * ```
+ *
+ * is transformed to:
+ * ```ts
+ * import {signal} from '@angular/core';
+ * const mySignal = signal('Hello World', ...(ngDevMode ? [{ debugName: "mySignal" }] : []));
+ * ```
+ *
+ * The transformer supports the following cases:
+ *
+ * # Variable declaration
+ * ```ts
+ * const mySignal = signal('Hello World');
+ * ```
+ *
+ * becomes
+ * ```
+ * const mySignal = signal(...(ngDevMode ? ['Hello World', { debugName: "mySignal" }] : ['Hello World']));
```

Comment:
Nit: I think we've now changed this to:

```suggestion
 * const mySignal = signal('Hello World', ...(ngDevMode ? [{ debugName: "mySignal" }] : []));
```